### Feature Engineering

In [22]:
import os, io, json, warnings
from datetime import datetime
warnings.filterwarnings("ignore")

import numpy as np
import pandas as pd
from dotenv import load_dotenv

import matplotlib.pyplot as plt
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score, accuracy_score, f1_score, roc_auc_score
from sklearn.linear_model import ElasticNet, LogisticRegression
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor, RandomForestClassifier
import joblib


In [23]:
path = Path.cwd().parent / "dataset" /"processed" /"preprocessed_data.csv"
print(path)

/Users/haseebsagheer/Documents/Python Learning/Cloud-Retail-Insights/dataset/processed/preprocessed_data.csv


In [24]:
df_base = pd.read_csv(path)
df = df_base
pd.set_option("display.max_column",None)

df.head()


OrderID   OrderDate    ShipDate        ShipMode CustomerID  \
0  ca-2017-152156  2017-11-08  2017-11-11    second class   cg-12520   
1  ca-2017-152156  2017-11-08  2017-11-11    second class   cg-12520   
2  ca-2017-138688  2017-06-12  2017-06-16    second class   dv-13045   
3  us-2016-108966  2016-10-11  2016-10-18  standard class   so-20335   
4  us-2016-108966  2016-10-11  2016-10-18  standard class   so-20335   

      CustomerName    Segment        Country             City       State  \
0      claire gute   consumer  united states        henderson    kentucky   
1      claire gute   consumer  united states        henderson    kentucky   
2  darrin van huff  corporate  united states      los angeles  california   
3   sean o'donnell   consumer  united states  fort lauderdale     florida   
4   sean o'donnell   consumer  united states  fort lauderdale     florida   

   PostalCode Region        ProductID         Category SubCategory  \
0       42420  south  FUR-BO-10001798        furniture   bookcases   
1       42420  south  FUR-CH-10000454        furniture      chairs   
2       90036   west  OFF-LA-10000240  office supplies      labels   
3       33311  south  FUR-TA-10000577        furniture      tables   
4       33311  south  OFF-ST-10000760  office supplies     storage   

                                         ProductName     Sales  DaysToShip  \
0                  bush somerset collection bookcase  261.9600           3   
1  hon deluxe fabric upholstered stacking chairs,...  731.9400           3   
2  self-adhesive address labels for typewriters b...   14.6200           4   
3      bretford cr4500 series slim rectangular table  957.5775           7   
4                     eldon fold 'n roll cart system   22.3680           7   

   OrderYear  OrderMonth  OrderQuarter OrderYearMonth  OrderWeekOfYear  \
0       2017          11             4        2017-11               45   
1       2017          11             4        2017-11               45   
2       2017           6             2        2017-06               24   
3       2016          10             4        2016-10               41   
4       2016          10             4        2016-10               41   

  OrderMonthName  OrderIsWeekendOrder  ShipYear  ShipMonth  ShipQuarter  \
0       November                False      2017         11            4   
1       November                False      2017         11            4   
2           June                False      2017          6            2   
3        October                False      2016         10            4   
4        October                False      2016         10            4   

  ShipYearMonth  ShipWeekOfYear ShipMonthName  ShipIsWeekendShip  
0       2017-11              45      November               True  
1       2017-11              45      November               True  
2       2017-06              24          June              False  
3       2016-10              42       October              False  
4       2016-10              42       October              False

In [25]:

pd.set_option("display.max_row",None)
print(df.columns)


Index(['OrderID', 'OrderDate', 'ShipDate', 'ShipMode', 'CustomerID',
       'CustomerName', 'Segment', 'Country', 'City', 'State', 'PostalCode',
       'Region', 'ProductID', 'Category', 'SubCategory', 'ProductName',
       'Sales', 'DaysToShip', 'OrderYear', 'OrderMonth', 'OrderQuarter',
       'OrderYearMonth', 'OrderWeekOfYear', 'OrderMonthName',
       'OrderIsWeekendOrder', 'ShipYear', 'ShipMonth', 'ShipQuarter',
       'ShipYearMonth', 'ShipWeekOfYear', 'ShipMonthName',
       'ShipIsWeekendShip'],
      dtype='object')


In [26]:

import os
import sys
from urllib.parse import quote_plus
from sqlalchemy import create_engine, text
from dotenv import load_dotenv
import pandas as pd
import pyodbc
import numpy as np
from pathlib import Path

DOTENV_PATH = "/Users/haseebsagheer/Documents/Python Learning/Cloud-Retail-Insights/secrets/.env"

# Force .env to override anything already in the process
if load_dotenv(dotenv_path=DOTENV_PATH, override=True):
    print("The .env file is loaded successfully.")
else:
    print("Warning: .env not found or could not be loaded.")

#This block of code will verify the SQL server login credentials
server   = os.getenv("SQL_SERVER")
database = os.getenv("SQL_DATABASE")
username = os.getenv("SQL_USERNAME")
password = os.getenv("SQL_PASSWORD")
print("Using SQL_SERVER   =", server)
print("Using SQL_DATABASE =", database)
print("Using SQL_USERNAME =", username)

if not all([server, database, username, password]):
    print("ERROR: Missing one or more of SQL_SERVER / SQL_DATABASE / SQL_USERNAME / SQL_PASSWORD")

DRIVER_PATH = "/opt/homebrew/lib/libmsodbcsql.18.dylib"

#Preparing the credentials for logging in the account (azure SQL Server)
odbc = (
    f"DRIVER={DRIVER_PATH};"
    f"SERVER={server};"
    f"DATABASE={database};"
    f"UID={username};"
    f"PWD={password};"
    "Encrypt=yes;"
    "TrustServerCertificate=no;"
    "Connection Timeout=30;"
)
print(odbc)

conn_url = f"mssql+pyodbc:///?odbc_connect={quote_plus(odbc)}"
print(conn_url)
try:
    engine = create_engine(conn_url, fast_executemany=True)
    print("SQLAlchemy engine created successfully.")
except Exception as e:
    print("Error creating engine:", e)


try:
    print("Testing connection...")
    with engine.connect() as conn:
        df = pd.read_sql("SELECT TOP 10 * FROM dbo.stg_sales;", conn)
        if df.empty:
            print("Query returned 0 rows.")
        else:
            print(f"Query returned {len(df)} rows")
            

except:
    print("There was something wrong in getting data from Azure SQl Server")






The .env file is loaded successfully.
Using SQL_SERVER   = sqlsrv-retail-dev.database.windows.net
Using SQL_DATABASE = sqldb-dretail-dev
Using SQL_USERNAME = sqladmin
DRIVER=/opt/homebrew/lib/libmsodbcsql.18.dylib;SERVER=sqlsrv-retail-dev.database.windows.net;DATABASE=sqldb-dretail-dev;UID=sqladmin;PWD=HASEEBSAGHEER12!@;Encrypt=yes;TrustServerCertificate=no;Connection Timeout=30;
mssql+pyodbc:///?odbc_connect=DRIVER%3D%2Fopt%2Fhomebrew%2Flib%2Flibmsodbcsql.18.dylib%3BSERVER%3Dsqlsrv-retail-dev.database.windows.net%3BDATABASE%3Dsqldb-dretail-dev%3BUID%3Dsqladmin%3BPWD%3DHASEEBSAGHEER12%21%40%3BEncrypt%3Dyes%3BTrustServerCertificate%3Dno%3BConnection+Timeout%3D30%3B
SQLAlchemy engine created successfully.
Testing connection...
Query returned 10 rows


In [27]:
# =========================
# Day 4: Local Modeling (No Azure writes)
# - Reuse YOUR working SQLAlchemy engine (no ODBC setup)
# - Read from SQL (or local fallback), train models, save artifacts locally
# =========================


# ----------------------------
# 0) Paths, folders, env
# ----------------------------
# Use your exact .env path (we won't rely on it for the engine, but it's handy to keep things tidy)
ENV_PATH = r"/Users/haseebsagheer/Documents/Python Learning/Cloud-Retail-Insights/secrets/.env"
load_dotenv(dotenv_path=ENV_PATH)

ROOT = os.getcwd()
ARTIF_DIR   = os.path.join(ROOT, "models")
REPORTS_DIR = os.path.join(ROOT, "reports")
INPUTS_DIR  = os.path.join(ROOT, "_inputs")   # optional local fallback
os.makedirs(ARTIF_DIR, exist_ok=True)
os.makedirs(REPORTS_DIR, exist_ok=True)
os.makedirs(INPUTS_DIR, exist_ok=True)

STAMP = datetime.utcnow().strftime("%Y%m%dT%H%M%SZ")



In [28]:
# ----------------------------
# 1) Create engine (PASTE YOUR WORKING CODE HERE)
# ----------------------------
# IMPORTANT:
# Paste the SAME engine creation code that worked in your “alternate method” script.
# Example placeholders (DO NOT use unless they are what you used and tested):
#   from sqlalchemy import create_engine
#   engine = create_engine("mssql+pytds://sqladmin:YOURPWD@sqlsrv-retail-dev.database.windows.net:1433/sqldb-retail-dev")
#
# Or whatever you used that successfully connected.
#
# After pasting, you should have: engine  (a working SQLAlchemy engine)

# === START: your engine code ===
# engine = <YOUR WORKING ENGINE HERE>
# === END: your engine code ===

try:
    if engine.name:
        print(f"Engine exist {engine.name}")
except NameError:
    raise RuntimeError("Please paste your working `engine = ...` code in the cell above before running.")

Engine exist mssql


In [29]:
def rmse(y_true, y_pred):
    return float(np.sqrt(mean_squared_error(y_true, y_pred)))

def read_from_sql_or_fallback():
    """
    Try reading from SQL using YOUR working engine.
    Prefer view dbo.v_processed_sales (if present), else fallback to fact_sales minimal cols.
    If SQL fails or returns empty, fallback to local files in ./_inputs/.
    """
    # 2a) Try SQL
    try:
        with engine.begin() as conn:
            has_view = False
            try:
                _ = pd.read_sql("SELECT TOP (1) * FROM dbo.facts_sales", conn)
                has_view = True
            except Exception:
                has_view = False

            if has_view:
                df = pd.read_sql("SELECT * FROM dbo.facts_sales", conn)
            else:
                q = """
                SELECT
                  f.OrderID, f.CustomerID, f.ProductID, f.Region,
                  f.OrderDate, f.ShipDate, f.Sales
                FROM dbo.fact_sales f
                """
                df = pd.read_sql(q, conn)
        if not df.empty:
            return df
    except Exception as e:
        print("SQL read failed, will try local fallback:", repr(e))

    # 2b) Fallback to local
    parquet_path = os.path.join(INPUTS_DIR, "processed_sales.parquet")
    csv_path     = os.path.join(INPUTS_DIR, "fact_sales.csv")

    if os.path.exists(parquet_path):
        print("Loading local:", parquet_path)
        return pd.read_parquet(parquet_path)
    if os.path.exists(csv_path):
        print("Loading local:", csv_path)
        return pd.read_csv(csv_path)

    raise FileNotFoundError(
        "Could not read from SQL and no local fallback found.\n"
        "Put a file at _inputs/processed_sales.parquet or _inputs/fact_sales.csv and rerun."
    )

# ----------------------------
# 3) Load data
# ----------------------------
df = read_from_sql_or_fallback()
print("Loaded shape:", df.shape)
print(df.head(3))


Loaded shape: (9788, 7)
          OrderID CustomerID        ProductID Region  OrderDate   ShipDate  \
0  ca-2017-152156   cg-12520  FUR-BO-10001798  south 2017-11-08 2017-11-11   
1  ca-2017-152156   cg-12520  FUR-CH-10000454  south 2017-11-08 2017-11-11   
2  ca-2017-138688   dv-13045  OFF-LA-10000240   west 2017-06-12 2017-06-16   

    Sales  
0  261.96  
1  731.94  
2   14.62  


In [30]:
def to_datetime_safe(s):
    return pd.to_datetime(s, errors='coerce', dayfirst=True)

if 'OrderDate' in df.columns and not np.issubdtype(df['OrderDate'].dtype, np.datetime64):
    df['OrderDate'] = to_datetime_safe(df['OrderDate'])
if 'ShipDate' in df.columns and not np.issubdtype(df['ShipDate'].dtype, np.datetime64):
    df['ShipDate'] = to_datetime_safe(df['ShipDate'])

# Derived from OrderDate
if 'OrderDate' in df.columns:
    df['OrderYear'] = df['OrderDate'].dt.year
    df['OrderMonth'] = df['OrderDate'].dt.month
    df['OrderQuarter'] = df['OrderDate'].dt.quarter
    df['OrderWeekOfYear'] = df['OrderDate'].dt.isocalendar().week.astype('Int64')
    df['OrderMonthName'] = df['OrderDate'].dt.month_name()
    df['OrderIsWeekendOrder'] = df['OrderDate'].dt.dayofweek.isin([5,6]).astype(int)
    df['OrderYearMonth'] = df['OrderDate'].dt.to_period('M').astype(str)

# Derived from ShipDate
if 'ShipDate' in df.columns:
    df['ShipYear'] = df['ShipDate'].dt.year
    df['ShipMonth'] = df['ShipDate'].dt.month
    df['ShipQuarter'] = df['ShipDate'].dt.quarter
    df['ShipWeekOfYear'] = df['ShipDate'].dt.isocalendar().week.astype('Int64')
    df['ShipMonthName'] = df['ShipDate'].dt.month_name()
    try:
        df['ShipIsWeekendShip'] = df['ShipDate'].dt.dayofweek.isin([5,6]).astype(int)
    except Exception:
        df['ShipIsWeekendShip'] = 0

# DaysToShip
if {'OrderDate','ShipDate'}.issubset(df.columns):
    df['DaysToShip'] = (df['ShipDate'] - df['OrderDate']).dt.days

# Clean strings for key categoricals
for col in ['ShipMode','Segment','Country','State','Region','Category','SubCategory','ProductName','CustomerName','City','PostalCode']:
    if col in df.columns and df[col].dtype == object:
        df[col] = df[col].astype(str).str.strip()

# Drop bad rows
need_cols = [c for c in ['OrderID','ProductID','CustomerID','OrderDate','Sales'] if c in df.columns]
df = df.dropna(subset=need_cols)
df['Sales'] = pd.to_numeric(df['Sales'], errors='coerce')
df = df.dropna(subset=['Sales'])
df = df[df['Sales'] >= 0]
print("Post-clean shape:", df.shape)


Post-clean shape: (9788, 21)


In [31]:
categorical_features = [
    'ShipMode','Segment','Country','State','Region','Category','SubCategory','OrderMonthName'
]
numeric_features = [
    'OrderYear','OrderMonth','OrderQuarter','OrderWeekOfYear','OrderIsWeekendOrder'
]
target_col = 'Sales'

drop_cols = [
    'CustomerName','City','PostalCode','ProductName',
    'OrderID','CustomerID','ProductID',
    'OrderDate','ShipDate',
    'ShipYear','ShipMonth','ShipQuarter','ShipYearMonth','ShipWeekOfYear','ShipMonthName','ShipIsWeekendShip','DaysToShip',
    'OrderYearMonth'
]
drop_cols = [c for c in drop_cols if c in df.columns]

X_all = df.drop(columns=[target_col] + drop_cols, errors='ignore')
y_all = df[target_col].astype(float)

# ----------------------------
# 6) Time-aware split
# ----------------------------
cutoff = df['OrderDate'].quantile(0.80)
train_mask = df['OrderDate'] <= cutoff
X_train, y_train = X_all[train_mask].copy(), y_all[train_mask].copy()
X_test,  y_test  = X_all[~train_mask].copy(), y_all[~train_mask].copy()

print("Cutoff:", cutoff.date(), "| Train:", X_train.shape, "| Test:", X_test.shape)

cat_cols = [c for c in categorical_features if c in X_train.columns]
num_cols = [c for c in numeric_features if c in X_train.columns]
print("Categoricals:", cat_cols)
print("Numerics:", num_cols)


Cutoff: 2018-07-20 | Train: (7831, 7) | Test: (1957, 7)
Categoricals: ['Region', 'OrderMonthName']
Numerics: ['OrderYear', 'OrderMonth', 'OrderQuarter', 'OrderWeekOfYear', 'OrderIsWeekendOrder']


In [32]:
preprocess = ColumnTransformer(
    transformers=[
        ("cat", OneHotEncoder(handle_unknown="ignore", sparse_output=True), cat_cols),
        ("num", Pipeline(steps=[("scale", StandardScaler())]), num_cols)
    ],
    remainder="drop"
)

# ----------------------------
# 8) Regression (ElasticNet, RF, GBR)
# ----------------------------
reg_models = {
    "elasticnet": ElasticNet(alpha=0.1, l1_ratio=0.5, random_state=42),
    "rf": RandomForestRegressor(n_estimators=300, n_jobs=-1, random_state=42),
    "gbr": GradientBoostingRegressor(random_state=42),
}
reg_results = {}
best_reg_name, best_reg_pipe, best_mae = None, None, 1e18

for name, est in reg_models.items():
    pipe = Pipeline(steps=[("prep", preprocess), ("model", est)])
    pipe.fit(X_train, y_train)
    preds = pipe.predict(X_test)
    mae  = mean_absolute_error(y_test, preds)
    mse  = mean_squared_error(y_test, preds)
    r2   = r2_score(y_test, preds)
    reg_results[name] = {"MAE": float(mae), "RMSE": float(np.sqrt(mse)), "R2": float(r2)}
    print(f"[REG {name}] MAE={mae:.2f} | RMSE={np.sqrt(mse):.2f} | R2={r2:.3f}")
    if mae < best_mae:
        best_mae, best_reg_name, best_reg_pipe = mae, name, pipe

print("\nBest regression:", best_reg_name, "MAE=", round(best_mae,2))

# Save regression artifacts
reg_model_path = os.path.join(ARTIF_DIR, f"regression_{best_reg_name}_{STAMP}.pkl")
reg_metrics_path = os.path.join(REPORTS_DIR, f"regression_metrics_{STAMP}.json")
joblib.dump(best_reg_pipe, reg_model_path)
with open(reg_metrics_path, "w") as f:
    json.dump({"best_model": best_reg_name, "results": reg_results, "cutoff": str(cutoff.date())}, f, indent=2)


[REG elasticnet] MAE=264.15 | RMSE=591.51 | R2=-0.002
[REG rf] MAE=290.38 | RMSE=624.65 | R2=-0.118
[REG gbr] MAE=279.88 | RMSE=594.36 | R2=-0.012

Best regression: elasticnet MAE= 264.15


In [33]:
threshold = y_train.quantile(0.80)
y_train_cls = (y_train >= threshold).astype(int)
y_test_cls  = (y_test  >= threshold).astype(int)

cls_models = {
    "logreg": LogisticRegression(max_iter=300),
    "rf": RandomForestClassifier(n_estimators=300, n_jobs=-1, class_weight="balanced", random_state=42),
}
cls_results = {}
best_cls_name, best_cls_pipe, best_f1 = None, None, -1

for name, est in cls_models.items():
    pipe = Pipeline(steps=[("prep", preprocess), ("model", est)])
    pipe.fit(X_train, y_train_cls)
    preds = pipe.predict(X_test)
    proba = pipe.predict_proba(X_test)[:,1] if hasattr(pipe, "predict_proba") else None
    acc = accuracy_score(y_test_cls, preds)
    f1  = f1_score(y_test_cls, preds)
    auc = roc_auc_score(y_test_cls, proba) if proba is not None else float("nan")
    cls_results[name] = {"ACC": float(acc), "F1": float(f1), "AUC": float(auc), "label_threshold": float(threshold)}
    print(f"[CLS {name}] ACC={acc:.3f} | F1={f1:.3f} | AUC={auc:.3f}")
    if f1 > best_f1:
        best_f1, best_cls_name, best_cls_pipe = f1, name, pipe

print("\nBest classifier:", best_cls_name, "F1=", round(best_f1,3))

# Save classification artifacts
cls_model_path = os.path.join(ARTIF_DIR, f"classifier_{best_cls_name}_{STAMP}.pkl")
cls_metrics_path = os.path.join(REPORTS_DIR, f"classifier_metrics_{STAMP}.json")
joblib.dump(best_cls_pipe, cls_model_path)
with open(cls_metrics_path, "w") as f:
    json.dump({"best_model": best_cls_name, "results": cls_results, "label_threshold": float(threshold)}, f, indent=2)


[CLS logreg] ACC=0.820 | F1=0.000 | AUC=0.508
[CLS rf] ACC=0.557 | F1=0.235 | AUC=0.490

Best classifier: rf F1= 0.235


In [34]:
preds = best_reg_pipe.predict(X_test)
plt.figure(figsize=(6,6))
plt.scatter(y_test, preds, alpha=0.35)
plt.plot([y_test.min(), y_test.max()], [y_test.min(), y_test.max()])
plt.xlabel("Actual Sales"); plt.ylabel("Predicted Sales")
plt.title("Actual vs Predicted (Test)")
plt.grid(True); plt.tight_layout()
plot1 = os.path.join(REPORTS_DIR, f"reg_actual_vs_pred_{STAMP}.png")
plt.savefig(plot1, dpi=140); plt.close()

with open(os.path.join(REPORTS_DIR, f"classification_threshold_{STAMP}.txt"), "w") as f:
    f.write(f"High-sale threshold (train 80th percentile): {threshold:.4f}\n")

print("\nSaved locally:")
print(" -", reg_model_path)
print(" -", reg_metrics_path)
print(" -", cls_model_path)
print(" -", cls_metrics_path)
print(" -", plot1)


Saved locally:
 - /Users/haseebsagheer/Documents/Python Learning/Cloud-Retail-Insights/notebooks/models/regression_elasticnet_20250818T112246Z.pkl
 - /Users/haseebsagheer/Documents/Python Learning/Cloud-Retail-Insights/notebooks/reports/regression_metrics_20250818T112246Z.json
 - /Users/haseebsagheer/Documents/Python Learning/Cloud-Retail-Insights/notebooks/models/classifier_rf_20250818T112246Z.pkl
 - /Users/haseebsagheer/Documents/Python Learning/Cloud-Retail-Insights/notebooks/reports/classifier_metrics_20250818T112246Z.json
 - /Users/haseebsagheer/Documents/Python Learning/Cloud-Retail-Insights/notebooks/reports/reg_actual_vs_pred_20250818T112246Z.png


## Day 4 Summary

- **Objective:** Train baseline models for forecasting (regression) and high-seller classification.
- **Data Used:** Cleaned sales dataset with derived time features (`OrderYear`, `OrderMonth`, `OrderQuarter`, etc.).
- **Models Tested:**
  - Regression: ElasticNet, Random Forest, Gradient Boosting
  - Classification: Logistic Regression, Random Forest
- **Key Results:**
  - Best regression → **ElasticNet** (lowest MAE ~264, but weak R²)
  - Best classifier → **Random Forest** (best F1 ~0.23, though still low)
- **Outputs Saved:** Metrics (JSON), trained models (Pickle), and plots (PNG).
- **Takeaway:** Baseline models trained successfully. Performance is limited with current features → next steps will focus on feature importance, engineering, and tuning.
